### PR TITLE
Provision Gateway during onboarding

### LINKED ISSUE #533: As a user, I want to redeem my invite code
As a user, I want to redeem my invite code, so I can get into the Testflight app.

### PR DESCRIPTION
**Pull request by bfollington**
Resolves https://github.com/subconsciousnetwork/subconscious/issues/468
Resolves #533 

# Images

Screenshot 2023-05-03 at 7 02 48 pm image image

Call out to `cloudctl` to attempt to provision a gateway from within the app.

- [x] Add field to capture invite code
- [x] Add retry option in settings
- [x] "Skip" or "Use offline" secondary button on First Run
- [x] Update & display provisioning status in settings
    - Different icons for different states, extract component?

### CODE REVIEW

File: xcode/Subconscious/Shared/Components/AppView.swift [hunk 1]
```diff
@@ -654,6 +697,43 @@ struct AppModel: ModelProtocol {
...
+            var model = state
+            model.gatewayURL = url
+            model.gatewayProvisioningStatus = .succeeded
+            return Update(state: model)
+            
+        case .failProvisionGateway(let error):
+            logger.error("Failed to provision gateway: \(error)")
+            var model = state
+            model.gatewayProvisioningStatus = .failed(error)
+            return Update(state: model)
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/AppView.swift:736]: Extract to helper method

**Review by gordonbrander — APPROVED** [review]
LGTM.

In future we may want to consider a word other than "provision" for user-facing copy. I'm not sure what yet exactly, but something familiar.

File: xcode/Subconscious/Shared/Services/GatewayProvisioningService.swift [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+                comment: "GatewayProvisioningService error description"
+            )
+        case .invalidStatusResponse(let res):
+            return String(
+                localized: """
+                    Invalid status response. \
+                    (status=\(res.status), \
+                    did=\(res.did ?? "<none>"), \
+                    url=\(res.url ?? "<none>"))
+                    """,
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Services/GatewayProvisioningService.swift:63]: TIL the closing `"""` determined the relative level of indentation captured within the string: https://www.swiftbysundell.com/tips/multiline-string-literal-tips-and-tricks/

File: xcode/Subconscious/Shared/Services/GatewayProvisioningService.swift [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+        } catch {
+            let attempts = attempts + 1
+            guard attempts < maxAttempts else {
+                return nil
+            }
+            
+            Self.logger.log("""
+            Waiting to re-check provisioning status, \
+            attempt \(attempts) of \(maxAttempts)
+            """)
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Services/GatewayProvisioningService.swift:190]: This seems to be the formatting suggested here: https://www.swiftbysundell.com/tips/multiline-string-literal-tips-and-tricks/

File: xcode/Subconscious/Shared/Services/GatewayProvisioningService.swift [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+            
+            sleep(UInt32(attempts))
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Services/GatewayProvisioningService.swift:192]: Simple linear backoff, we could easily switch to exponential if this is too eager.

Comment (bfollington) [reply xcode/Subconscious/Shared/Services/GatewayProvisioningService.swift:192]: Note: extract this waiting loop somewhere else, we will want to re-use it.

Do 2^n with a configurable maximum wait time.

**Pull request merged**